### ISSUE TITLE
All rendering has jagged edges on iOS while recording or taking screenshots

### ISSUE THREAD

**Issue by PabloNeirotti** [godotengine/godot#82730]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

iPhone 13 Pro Max – iOS 17.0.2

### Issue description

All rendering on screen, 2D or 3D, has jagged edges on iPhone screenshots or recordings, but not when no recording is taking place.

The issue is present on:

1. **Screenshots of a Godot game** (using the screenshot key combinations on the iPhone)
2. **Recordings of a Godot game** (recorded using the iPhone native record-screen action found at the control center). _Both the output video and the screen, while recording, show jagged edges on a Godot game. Once recording stops, jagged edges disappear, but of course remain on the recorded file._

Just to be clear, the image _seems_ to be fine when no recording is taking place, or when not looking at a screenshot. I don't know if the rendering scale of Godot changes when a recording or screenshot take place, **or whether there is an upscaling smoothing that goes away when these recordings or screenshots take place.**

---

### Could it be stretching? I don't think so.

At first I thought this was related to stretching or HiDPI settings, so I played a lot with that, to no avail. Here is an example of how it looks on the MRP:

<img width="660" alt="Screenshot 2023-10-03 at 3 45 58 PM" src="https://github.com/godotengine/godot/assets/905216/798d1895-151a-478d-b83f-4d07f3e1876e">

But of course, **the issue only exists while recording or on screenshots,** so it wouldn't make sense that is related to this. Why aren't edges always jagged?

The jagginess we see in screenshots/recordings is exactly you'd see if you are indeed using **Stretch** for the window and have the aspect expand.

---

### Visual comparisons

It's easier to see in person. I've made a comparison between a screenshot and a photo taken with another phone. Of course this is not ideal, as the camera introduces it's own distortions and blurriness, but I think it helps to showcase the difference quite well.

1. Look at the arrows.
2. Look at the eyes of our little Godot buddy.

**ZOOM IN A LOT**

![Jagged Edges](https://github.com/godotengine/godot/assets/905216/280ecafb-e93d-4aa2-b683-8c621ec3c803)

---

### Conclusion

The main issue with this is that any kind of marketing asset you want to make out of the game might require recording outside the target device, which is not ideal. Or anyone wanting to share a screenshot of the game will make it look ugly, which again is pretty bad presentation-wise.

I have not noticed any other app or game I have do the same on my iPhone.

If you need any more information, please let me know and I'll do my best to provide it.

### Steps to reproduce

1. New project with any 3D objects, some UI elements as well (or use included MRP).
2. Export to iOS, install on iPhone.
3. Start a recording and see how all edges become jagged, real time. Video will include jagged edges as well. Alternatively, take a screenshot and zoom in to see jaggies better. Otherwise edges will look fine.

### Minimal reproduction project

Any fresh project will have this issue, but I am sharing one that I also tested on iPhone.

[Godot Bugs Playground.zip](https://github.com/godotengine/godot/files/12793073/Godot.Bugs.Playground.zip)

**Comment by PabloNeirotti** [comment]
Additional note: I think this might be related to the "window size" of Godot not matching the screen size of iPhone.

The iPhone 13 Pro Max, which I'm testing on, is specified to have a display of `2778‑by‑1284-pixel resolution at 458 ppi`

However when I `print(get_window().size)`, on the Xcode console (while running on my iPhone) I see: `(1125, 2436)`

So this could explain why the screen looks a bit fuzzy, and screenshots maybe simply loose a filtering that made things smooth as opposed to "nearest neighbor".

In other words, it'd seem that Godot is running at 87.6% of the native screen resolution. I am trying to see if there is any way I can force the size this via scripts but so far it doesn't look promising.

Not sure how device resolution is detected, but maybe that's where the issue lays on.

**Comment by akien-mga** [comment]
CC @bruvzg

**Comment by PabloNeirotti** [comment]
Any plans to use the native resolution on iOS?

Not only it's blurry, but the sizes are all wrong since the entire viewport is being scaled down and stretched across the screen. Makes it really difficult to test the UI and get a real feel of how big buttons should be on screen for a user to tap.

**Comment by Calinou** [comment]
> Any plans to use the native resolution on iOS?

This is likely a bug, not the result of an intentional decision (especially since a scale factor of ~87.6% doesn't translate to anything meaningful).

**Comment by bruvzg** [comment]
Most likely related to `screen_get_scale`, and probably only relevant Pro Max iPhone models. `screen_get_scale` was changed multiple times to resolve the similar issue (https://github.com/godotengine/godot/pull/70478), so I'm not sure what's the correct way to do it.

**Comment by bruvzg** [comment]
Also, it might be exactly how it should work. At least on macOS, if you enable hiDPI mode on 4K display, it will pretend to have 5120×2880 pixels instead of real 3840×2160.

**Comment by PabloNeirotti** [comment]
Thank you for your response. If you need help testing this, let me know.

That said, I don't agree that this is how it should work. Any other app or game I have on my iPhone has crisp graphics. Unity or Unreal games work just fine, only Godot can't reach the full screen resolution. Regardless if something is "technically correct" code-wise, I think we can agree the result is undesirable. No one likes to read blurry text, or see smudgy edges.

**Comment by bruvzg** [comment]
Well this is even stranger, I can reproduce it on iOS simulator for 13 Pro Max. Seems like all iOS APIs return this values, and it depends on whether the project is using LaunchScreen storyboard or LacnchImages (but in both cases it's wrong).
1125x2436 and 1179x2556 instead of 1284x2778.

**Comment by bruvzg** [comment]
Default export config should be fixed by https://github.com/godotengine/godot/pull/86312, and it should be fixable in the existing versions by checking `Use launch screen storyboard` in the export settings and re-exporting the project.
